stan/math/prim/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,64 @@
+#ifndef STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/chol2inv.hpp>
+#include <stan/math/prim/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/quad_form.hpp>
+#include <stan/math/prim/fun/inverse_spd.hpp>
+#include <stan/math/prim/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/mdivide_right_spd.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the Moore-Penrose generalized inverse of the specified matrix.
+ *
+ * @tparam T type of elements in the matrix
+ * @tparam n number of rows, can be Eigen::Dynamic
+ * @tparam m number of columns, can be Eigen::Dynamic
+ *
+ * @param G specified matrix
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ */
+template <typename EigMat,
+          require_eigen_vt<std::is_arithmetic, EigMat>* = nullptr>
+inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
+                     EigMat::ColsAtCompileTime>
+generalized_inverse(const EigMat& G) {
+  using value_t = value_type_t<EigMat>;
+  if (G.size() == 0) {
+    return {};
+  }
+
+  const auto n = G.rows();
+  const auto m = G.cols();
+
+  if (G.rows() == G.cols()) {
+    return G.inverse();
```

Comment:
I think it would be better to call stan's inverse function here so that if the matrix has vars in it then it will use the reverse mode specialization

Reply:
Done